### ISSUE TITLE
Bundling joi with dynamicRequireTargets option

### ISSUE THREAD

**Issue by perrin4869** [rollup/plugins#731]
- Rollup Plugin Name: @rollup/plugin-commonjs
- Rollup Plugin Version: 17.0.0
- Rollup Version: 2.35.1
- Operating System (or Browser): slackware64-current
- Node Version: 14.6
- Link to reproduction _(⚠️ read below)_: https://github.com/perrin4869/rollup-joi

### Expected Behavior

```
❯ npm start

> rollup-readable-stream@1.0.0 start /home/perrin4869/rollup-joi
> node index.out.js

{
  value: { name: 1 },
  error: [Error [ValidationError]: "name" must be a string] {
    _original: { name: 1 },
    details: [ [Object] ]
  }
}
```

### Actual Behavior

```
❯ npm start

> rollup-readable-stream@1.0.0 start /home/perrin4869/rollup-joi
> node index.out.js

/home/perrin4869/univapay/rollup-joi/index.out.js:14598
  name: lib.string()
            ^

TypeError: lib.string is not a function
    at Object.<anonymous> (/home/perrin4869/univapay/rollup-joi/index.out.js:14598:13)
    at Module._compile (internal/modules/cjs/loader.js:1236:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1257:10)
    at Module.load (internal/modules/cjs/loader.js:1085:32)
    at Function.Module._load (internal/modules/cjs/loader.js:950:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)
    at internal/main/run_main_module.js:17:47
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! rollup-readable-stream@1.0.0 start: `node index.out.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the rollup-readable-stream@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/perrin4869/.npm/_logs/2020-12-14T17_17_31_946Z-debug.log
```

### Additional Information

Related to https://github.com/rollup/plugins/issues/545
`joi` is a cjs module that contains circular dependencies (https://github.com/sideway/joi/issues/2517).
I am trying to bundle it with rollup, but when importing it with `import Joi from "joi"`, it will bundle incorrectly, resulting in an error.
When using `const Joi = require("joi")` it will bundle correctly, however, in my actual project I am using `typescript`, therefore it is not an option to use `require`. I think both options should work though.

**Comment by shellscape** [comment]
_(For what it's worth, `joi` is a mess of a project and the hapi community is heavily fragmented around it. I'd recommend giving `superstruct` or `yup` a look, both of which play nicely with Rollup)_

**Comment by perrin4869** [comment]
`superstruct` I checked but is miles behind Joi in terms of features, it doesn't fulfill my requirements. `yup` gets much closer, but still seems to be missing key features from joi...

**Comment by shellscape** [comment]
That's interesting. I found joi lacking compared to the two. Not to get too far off topic, but I'm genuinely curious to know how the two come up short.

**Comment by perrin4869** [comment]
I like about `superstruct` the basic API, based on functions rather than method chaining... however doesn't have a way to run async validation, or the `alter`/`tailor` functionality from `joi`. `yup` is also missing `alter`/`tailor`, which is incredibly useful writing a single schema that can validate `inserts`/`updates`/`normal`.

**Comment by perrin4869** [comment]
I think `yup` doesn't support validating objects by [`patterns`](https://joi.dev/api/?v=17.3.0#objectpatternpattern-schema-options) either... I would need to try to rewrite my schemas to have a more clear comparison though.
I do agree that Joi seems to be more messy, but it has invaluable features that are actually useful

**Comment by shellscape** [comment]
@ianstormtaylor I'd be curious on your take here.

@perrin4869 sorry for the side track, but it might help the community. we'll see about looking at the issue you're facing with commonjs as soon as possible.

**Comment by perrin4869** [comment]
Of course I understand! @shellscape 
Thanks!

**Comment by perrin4869** [comment]
@danielgindi Sorry to keep bothering you, but since you wrote the `dynamicRequireTargets` feature, you can give the best advice here!
I think this is a very important feature of this plugin and could use further examples/explanations of how to get right. As of right now I couldn't get it to work ><;;
Let me link this issue: https://github.com/rollup/plugins/issues/545

**Comment by danielgindi** [comment]
I actually recommend `ajv` as it compiles standard JSON Schema into a validator function, and is much faster. Sometimes 4000x faster as I measured in one project lately.

Anyway, the commonjs plugin will transform `require` calls to dynamic targets into a wrapped reference to a bundle of the module (emulated to load exactly as in commonjs).
The thing is if it's a simple `import` then - I'm not sure if commonjs plugin transforms it.
Anyway I'd try with `transformMixedEsModules` set to true, and also set `extensions` to also include your extensions (ts, vue, whatever they are).
This *must* be done as transpiling commonjs stuff is the unique responsibility of the `commonjs` plugin,

**Comment by perrin4869** [comment]
Actually, I setup a bare `js` project and tried `transformMixedEsModules: true`, etc, here: https://github.com/perrin4869/rollup-joi/blob/master/rollup.config.js
Of course couldn't get it to work, otherwise I wouldn't be asking ><;;
I'll be forever indebted if you could take a look at it ^^;;

**Comment by shellscape** [comment]
@danielgindi give superstruct a look. tinier and faster than ajv even.

**Comment by danielgindi** [comment]
> @danielgindi give superstruct a look. tinier and faster than ajv even.

I like working with a standard JSON schema,  this allows me to easily switch to binary encoding (i.e translating to `avsc` or other means).

**Comment by danielgindi** [comment]
> Actually, I setup a bare `js` project and tried `transformMixedEsModules: true`, etc, here: https://github.com/perrin4869/rollup-joi/blob/master/rollup.config.js
> Of course couldn't get it to work, otherwise I wouldn't be asking ><;;
> I'll be forever indebted if you could take a look at it ^^;;

1) The easiest method would be to use `@rollup/plugin-alias` and alias joi into `node_modules/joi/dist/joi-browser.min.js`.

2) If you do package the joi as commonjs, you should use `"**/node_modules/joi/lib/**/*.js"`, to not include the `dist` part of it. As it inflates your package.

3) There's a chance I could solve this for ES. By returning a `commonjsRequire` call from the 'load' function, and rewrite the actual loading path in normal `require` paths to a proxied path.

4) A temporary workaround would be to keep a local `joi-proxy.js` with `const Joi = require('joi'); export default Joi;`, then use `import Joi from 'joi-proxy'`.

**Comment by danielgindi** [comment]
Rethinking (3) - I'm not sure if it's currently possible. As if I would do this and the modules are actually loaded from some kind of a virtual loader - it would break (no way I'm aware of to call the default loader from within the `load` callback with another path).  @shellscape correct me if I'm wrong?

**Comment by danielgindi** [comment]
> Rethinking (3) - I'm not sure if it's currently possible. As if I would do this and the modules are actually loaded from some kind of a virtual loader - it would break (no way I'm aware of to call the default loader from within the `load` callback with another path). @shellscape correct me if I'm wrong?

Answering my own question - we are already loading the dynamic files this way anyway in order to wrap them. So no problem there.

**Comment by perrin4869** [comment]
Ah glad that you mentioned (1), I did try that way back when I opened this issue, but `joi-browser.min.js` was built in such a way that it requires a global `window` object (otherwise it throws when loading the module), and I am running this in a serverless environment... so not ideal, since I would also need to add a polyfill.
I'll definitely take (2) to heart!
Number (4) worked like a charm!!!! Thank you so much!!!!!!! Highly anticipating (3) :D

**Comment by danielgindi** [comment]
@perrin4869 Glad I helped :-)

About (2) - I'd love it if pacakge owners documented on their pages what is the correct formula for dynamic require with the least files to package. Then again, I'd love it if they refactored they packages to not need that at all... So I guess that having a document in the rollup realm would work. @shellscape what do you say?

About (3) - A PR is pending :-)

**Comment by perrin4869** [comment]
closed by #766 
@danielgindi אני חייב לך בירה כשאבקר בישראל מתישהו 😄

**Comment by danielgindi** [comment]
> closed by #766
> @danielgindi אני חייב לך בירה כשאבקר בישראל מתישהו 😄

Let's do this 🍻

**Comment by perrin4869** [comment]
@shellscape Any clue when you can release this feature?
